src/conventions/h-group/take-action.js

```diff
@@ -263,10 +259,33 @@ export function take_action(game) {
 		}
 	}
 
+	/** @param {Clue} clue */
+	const not_selfish = (clue) => {
+		const list = state.hands[clue.target].clueTouched(clue, state.variant).map(c => c.order);
+		const { focused_card } = determine_focus(state.hands[clue.target], common, list, { beforeClue: true });
+		const { suitIndex } = focused_card;
+
+		return common.hypo_stacks[suitIndex] === state.play_stacks[suitIndex] ||
+			Utils.range(state.play_stacks[suitIndex] + 1, common.hypo_stacks[suitIndex] + 1).every(rank =>
+				!state.hands[state.ourPlayerIndex].some(c => me.thoughts[c.order].matches({ suitIndex, rank })));
+	};
+
+	// Consider finesses while finessed if we are only waited on to play one card,
+	// it's not a selfish finesse, doesn't require more than one play from our own hand,
+	// and we're not in the end-game.
+	const waiting_self_connections = game.common.waiting_connections.filter(c => c.connections[0]?.reacting === state.ourPlayerIndex);
+	const waiting_cards = waiting_self_connections.reduce((sum, conn) => sum + conn.connections.length, 0);
+	const waiting_out_of_order = waiting_self_connections.some(c => c.connections.length >= 2 && !inBetween(state.numPlayers, c.connections[1].reacting, state.ourPlayerIndex, c.connections[2]?.reacting ?? c.target));
```

Comment:
This is a lot closer now, thanks. :+1: Just one more thing, the waiting connections might be in progress by our turn (e.g. we are part of an r4 finesse, someone already played r1 and we have r2 and would be holding up r3 and r4). This is tracked by `WaitingConnection.conn_index`, which is the index of the connection that should occur next. The connections aren't popped from the array as they occur, only `conn_index` is incremented. Thus, these conditions need to use `conn_index` as the starting point rather than 0.

Why did I do it this way? Not sure, really. :P

Reply:
Oh I had no idea they kept the already played parts of the connection! I'm sure you would have caught if I misused the connections anywhere else, your reviews have all been very thorough :-).

Updated.